### DISCUSSION TITLE
How to include local OSGI Bundles

### DISCUSSION THREAD

**Discussion by fkellner** [eclipse-tycho/tycho#3539] [Q&A]
Hello!

I have recently been onboarded to a project that has been using the Eclipse export wizard for an RCP product, but we want to migrate to using Tycho builds. So far I followed the [demo tutorial](https://github.com/jsievers/tycho-demo) and, apart from needing to update the version and the p2 site url, step 1 seems to work fine. The only problem is that the product requires OSGI bundles from neighbouring folders:
```
# Project Structure:
- jarprovider
    - src
    - ...
    - META-INF
      - MANIFEST.MF
- model
    - src
    - ...
    - META-INF
      - MANIFEST.MF
- workbench
    - src
    - ...
    - META-INF
      - MANIFEST.MF
    - workbench.product
    - plugin.xml
    - pom.xml
```
Content of the pom.xml:
```xml
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <groupId>REDACTED</groupId>
  <artifactId>workbench</artifactId>
  <version>1.0.0-SNAPSHOT</version>
  <packaging>eclipse-plugin</packaging>
  <properties>
    <tycho-version>4.0.6</tycho-version>
  </properties>
  <build>
    <plugins>
      <plugin>
        <!-- enable the Tycho build extension -->
        <groupId>org.eclipse.tycho</groupId>
        <artifactId>tycho-maven-plugin</artifactId>
        <version>${tycho-version}</version>
        <extensions>true</extensions>
      </plugin>
    </plugins>
  </build>
  <repositories>
        <repository>
		      <id>eclipse-project-updates</id>
		      <layout>p2</layout>
		      <url>http://download.eclipse.org/eclipse/updates/4.30</url>
		</repository>
		<repository>
		      <id>eclipse-project-2022-06</id>
		      <layout>p2</layout>
		      <url>https://download.eclipse.org/releases/2022-06</url>
		</repository>
    </repositories>
</project>
```

```mf
...
# MANIFEST.MF
Require-Bundle: org.eclipse.ui,
 org.eclipse.core.runtime,
 org.eclipse.draw2d;bundle-version="3.8.1",
 model;bundle-version="1.0.0",
 org.eclipse.jface,
 jarprovider;bundle-version="1.0.0"
...
```
When running the Tycho build, I get:
```
[[1;31mERROR[m] Cannot resolve dependencies of project <redacted>:workbench:eclipse-plugin:1.0.0-SNAPSHOT
[[1;31mERROR[m]  with context {osgi.os=win32, osgi.ws=win32, org.eclipse.update.install.features=true, osgi.arch=x86_64, org.eclipse.update.install.sources=true}
[[1;31mERROR[m]   Software being installed: workbench 1.0.0.qualifier
[[1;31mERROR[m]   Missing requirement: workbench 1.0.0.qualifier requires 'osgi.bundle; model 1.0.0' but it could not be found: See log for details
```
Those folders/modules/whatever are not Maven projects yet. 

Reading the available resources, it seems like Tycho needs to resolve those dependencies from p2 repositories, and for local bundles, there is the possibility to create a filesystem p2 repository. 

Can you please outline to me the steps that are necessary to get this to build? The available resources all seem to have different use cases and it is hard for me to discern which configuration setting results in what exactly happening.

Despite having invested some time already, I am not fully proficient in the domain language of products, features, plugins etc. as of yet. Resources that help me get a better understanding of those would also be very welcome.

Thanks in Advance!

**Comment by phermsdorf** [comment]
For tycho to find your other bundles they also need to be part of the maven build.
In your linked demo project you can see that the bundle also has a reference to the parent pom and in the parent pom the bundle is listed as module. 

That seems to be missing from your build.

Another way to achive the same, but without explicitly defining pom files for all parts of your build is to setup a pomless tycho build.  See here for more information https://github.com/eclipse-tycho/tycho/wiki/Tycho-Pomless

Bye Peter

**Comment by laeubi** [comment]
I would recommend to choose a structured build layout (your example already almost looks like that):

https://tycho.eclipseprojects.io/doc/latest/StructuredBuild.html

you can of course list all modules manually as well, but as mentioned by @phermsdorf this usually happens in one parent pom.

**Comment by fkellner** [comment]
Thank you for your replies! I have tried converting to a structured build layout, by converting it as follows:
```
# Project Structure:
- bundles
  - jarprovider
      - src
      - ...
      - META-INF
        - MANIFEST.MF
  - model
      - src
      - ...
      - META-INF
        - MANIFEST.MF
- products
  - workbench
      - src
      - ...
      - META-INF
        - MANIFEST.MF
      - workbench.product
      - plugin.xml
- pom.xml
- target-platform.target
```

My `pom.xml` now looks like this:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd"
	xmlns="http://maven.apache.org/POM/4.0.0"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
	<modelVersion>4.0.0</modelVersion>
	<groupId>...</groupId>
	<artifactId>parent</artifactId>
	<version>1.0.0-SNAPSHOT</version>
	<packaging>pom</packaging>
	
	<modules>
		<module>bundles</module>
		<!--<module>features</module>
		<module>sites</module>-->
		<module>products</module>
	</modules>
	
	<build>
		<plugins>
			<plugin>
				<groupId>org.eclipse.tycho</groupId>
				<artifactId>tycho-maven-plugin</artifactId>
				<version>${tycho-version}</version>
				<extensions>true</extensions>
			</plugin>
			<plugin>
				<groupId>org.eclipse.tycho</groupId>
				<artifactId>target-platform-configuration</artifactId>
				<version>${tycho-version}</version>
				<configuration>
					<environments>
					 <!--<environment>
					   <os>linux</os>
					   <ws>gtk</ws>
					   <arch>x86_64</arch>
					 </environment>
					 <environment>
					   <os>linux</os>
					   <ws>gtk</ws>
					   <arch>ppc64le</arch>
					 </environment>
					  <environment>
					   <os>linux</os>
					   <ws>gtk</ws>
					   <arch>aarch64</arch>
					 </environment>-->
					 <environment>
					   <os>win32</os>
					   <ws>win32</ws>
					   <arch>x86_64</arch>
					 </environment>
					 <!--<environment>
					   <os>macosx</os>
					   <ws>cocoa</ws>
					   <arch>x86_64</arch>
					 </environment>
					 <environment>
					   <os>macosx</os>
					   <ws>cocoa</ws>
					   <arch>aarch64</arch>
					 </environment>-->
					</environments>
				</configuration>
			</plugin>
		</plugins>
	</build>
	
	<repositories>
		<repository>
		      <id>eclipse-project-updates</id>
		      <layout>p2</layout>
		      <url>http://download.eclipse.org/eclipse/updates/4.30</url>
		</repository>
		<repository>
		      <id>eclipse-project-2022-06</id>
		      <layout>p2</layout>
		      <url>https://download.eclipse.org/releases/2022-06</url>
		</repository>
    </repositories>
</project>

```

`mvn clean verify` now runs fine and produces `.jar`-files in the appropriate `target` folders.

Now, again, please forgive me for asking a beginner question, but how or where would I configure to get an actual export that includes an executable launcher for the product? I had a look at the [Packaging Types](https://tycho.eclipseprojects.io/doc/latest/PackagingTypes.html), but I only found options for plugins and features, but not for a [PDE Product](https://wiki.eclipse.org/PDE/User_Guide#Product). 
I understand that apparently for legal reasons, the export cannot contain a JRE, however, the GUI export provides me with a folder that includes eclipse as well as the exported plugins and a launcher as configured in the `.product`-file, and one only needs to add a JRE.

Edit: As per the [Tutorial Section 5](https://github.com/jsievers/tycho-demo/tree/master/exercises/Exercise_05_Add_Product/tychodemo.parent), apparently this is the job of the [Tycho P2 Director Plugin](https://tycho.eclipseprojects.io/doc/latest/tycho-p2-director-plugin/plugin-info.html)

**Reply by laeubi** [reply]
Your `workbench.product` must not be contained within a plugin (what it currently is), so move your "workbench" to the plugins, and the `workbench.product` into products/product/workbench.product for example.

**Reply by fkellner** [reply]
Ah, okay, thank you! After changing this, I get a new error: 
```
[ERROR] Cannot resolve dependencies of project...:workbench.product:eclipse-repository:3.0.0
[ERROR]  with context {osgi.os=win32, osgi.ws=win32, org.eclipse.update.install.features=true, osgi.arch=x86_64, org.eclipse.update.install.sources=true}
[ERROR]   Software being installed: workbench.product 3.0.0
[ERROR]   Missing requirement: workbench.product 3.0.0 requires 'org.eclipse.equinox.p2.iu; org.eclipse.equinox.util 0.0.0' but it could not be found: See log for details
```
I do understand that I may have to provide a new repository which contains the missing bundles, but half an hour of googling has left me stumped - is there any meta search where you can type in a bundle and get the repository containing it that I am not aware of?

**Reply by laeubi** [reply]
Basically you should have a `.target` file or you need to use the updatesite of the eclipse you are running as described here:

https://tycho.eclipseprojects.io/doc/latest/TargetPlatform.html

**Reply by fkellner** [reply]
Yes, well I read that far, and it states in the [Simple target platform configuration](https://tycho.eclipseprojects.io/doc/latest/TargetPlatform.html#simple-target-platform-configuration):

> In terms of the target platform, this means that the entire content of the specified p2 repositories become part of the target platform.

And I added the p2 repository of my current distribution as such:
```xml
<repository>
  <id>eclipse-2023-03</id>
  <layout>p2</layout>
  <url>https://download.eclipse.org/releases/2023-03</url>
</repository>
```
So the fact that I get the error that `org.eclipse.equinox.p2.iu` and `org.eclipse.equinox.util` are not found _after I added the repositories_ suggests that these bundles are not in there, right? 

I have tried the following repositories:
- http://download.eclipse.org/eclipse/updates/4.30
- https://download.eclipse.org/releases/2022-06
- https://download.eclipse.org/releases/latest
- https://download.eclipse.org/releases/2023-03
- http://download.eclipse.org/tools/orbit/downloads/drops/R20160520211859/repository/ (from a StackOverflow answer)
- http://download.oracle.com/otn_software/oepe/library/eclipse-oxygen.2 (also from a StackOverflow answer)

**Reply by fkellner** [reply]
Moving this to a `platform-target.target`-file referenced by the POM does not solve the problem, but results in the same error.

**Reply by laeubi** [reply]
In general it not works to try random things, your snippet looks correct if you are targeting that release, but its most often easier to show the problem with an example project. Its also not completely true what you are using inside PDE and how your product looks like and what tycho version you are using.

**Reply by fkellner** [reply]
Sorry, here is the `.mvn/maven.config` (as suggested by the tutorial):
```
-Dtycho-version=4.0.6
```
I would need to ask with my employer if I can provide more source code. If I find a solution with my colleagues, I will post it here, too!

**Reply by laeubi** [reply]
You can always try to create a small example project, that often shows already the problem.

**Reply by fkellner** [reply]
https://www.eclipse.org/forums/index.php?t=msg&th=1108070/
Problem is still not solved, but I seem to not be the first one stumbling there.
Apparently, the `org.eclipse.equinox.p2.iu:` in the error message denotes the type of bundle that is missing (i.e. in this case an "Installable Unit"). The person got the problem solved, but did not say how, so I will ask there.
